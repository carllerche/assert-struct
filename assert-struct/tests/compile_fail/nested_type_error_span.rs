use assert_struct::assert_struct;

#[derive(Debug)]
struct QuerySql {
    stmt: Statement,
    ret: Option<String>,
}

#[derive(Debug)]
enum Statement {
    Query(QueryStmt),
}

#[derive(Debug)]
struct QueryStmt {
    body: ExprSet,
}

#[derive(Debug)]
enum ExprSet {
    Select(SelectStmt),
}

#[derive(Debug)]
struct SelectStmt {
    source: Source,
    filter: Expr,
}

#[derive(Debug)]
enum Source {
    Table(Vec<TableRef>),
}

#[derive(Debug)]
struct TableRef {
    table: String,
}

#[derive(Debug)]
enum Expr {
    BinaryOp(BinaryOpExpr),
}

#[derive(Debug)]
struct BinaryOpExpr {
    lhs: Box<String>,
    op: BinaryOp,
    rhs: Box<String>,
}

#[derive(Debug)]
enum BinaryOp {
    Eq,
}

fn main() {
    let query_sql = QuerySql {
        stmt: Statement::Query(QueryStmt {
            body: ExprSet::Select(SelectStmt {
                source: Source::Table(vec![TableRef { table: "users".to_string() }]),
                filter: Expr::BinaryOp(BinaryOpExpr {
                    lhs: Box::new("col".to_string()),
                    op: BinaryOp::Eq,
                    rhs: Box::new("val".to_string()),
                }),
            }),
        }),
        ret: Some("result".to_string()),
    };

    assert_struct!(query_sql, _ {
        stmt: Statement::Query(_ {
            body: ExprSet::Select(_ {
                source: Source::Table([
                    _ { table: "users", .. },
                ]),
                filter: Expr::BinaryOp(_ {
                    *lhs: "col",
                    op: BinaryOp::Eq,
                    *rhs: == 123,  // Type error: can't compare String with integer
                    ..
                }),
                ..
            }),
            ..
        }),
        ret: Some(_),
        ..
    });
}