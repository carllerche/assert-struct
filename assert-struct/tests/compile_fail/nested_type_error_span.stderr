error[E0277]: can't compare `String` with `{integer}`
  --> tests/compile_fail/nested_type_error_span.rs:81:30
   |
81 |                     *rhs: == 123,  // Type error: can't compare String with integer
   |                              ^^^ no implementation for `String == {integer}`
   |
   = help: the trait `PartialEq<{integer}>` is not implemented for `String`
   = help: the following other types implement trait `PartialEq<Rhs>`:
             `String` implements `PartialEq<&str>`
             `String` implements `PartialEq<ByteStr>`
             `String` implements `PartialEq<ByteString>`
             `String` implements `PartialEq<Cow<'_, str>>`
             `String` implements `PartialEq<str>`
             `String` implements `PartialEq`
   = note: required for `&String` to implement `PartialEq<&{integer}>`
